lib/beautiful/colorful.lua

```diff
@@ -0,0 +1,1395 @@
+        -- do we have red/green/blue/alpha fields?
+        if val.red ~= nil and val.green ~= nil and val.blue ~= nil then
+            local alpha = val.alpha or 255
+            return color.from.rgba(val.red, val.green, val.blue, alpha)
+        -- do we have hue/saturation/lightness/alpha fields?
+        elseif val.hue ~= nil and val.saturation ~= nil and val.lightness ~= nil then
+            local alpha = val.alpha or 1
+            return color.from.hsla(val.hue, val.saturation, val.lightness, alpha)
+        -- do we have a table of 3 or 4 values?
+        elseif #val == 3 or #val == 4 then
+            -- if val[1] [0,360] and val[2] and val[3] [0,1] we deduce HSL
+            if (between(val[1], 0, 360) and
+                between(val[2], 0, 1)   and
+                between(val[3], 0, 1)      )
+            then
+                local alpha = val[4] and clamp(val[4], 0, 1) or 1
+                return color.from.hsla(
+                    val[1],
+                    val[2],
+                    val[3],
+                    alpha
+                )
+            end
+            if (between(val[1], 0, 255) and
+                between(val[2], 0, 255) and
+                between(val[3], 0, 255)    )
+            then
+                local alpha = val[4] and clamp(val[4], 0, 255) or 255
+                return color.from.rgba(
+                    val[1],
+                    val[2],
+                    val[3],
+                    alpha
+                )
+            end
+            -- if values aren't in range, escape
+        end
+        -- don't know the table type, escape
+    end
+    -- check for C userdata (GdkRGBA) with values
+    if type(val) == "userdata" then
+        if val.red ~= nil and val.green ~= nil and val.blue ~= nil then
+            local alpha = val.alpha or 1
+            return color.from.rgba(
+                val.red * 255,
+                val.green * 255,
+                val.blue * 255,
+                alpha * 255
+            )
+        end
+    end
+    -- not a valid type, quit
+    return nil
+end
+
+-- }}} color.new()
+
+-- {{{ String
+-- {{{ Helper tables and functions
+
+local css = {
+    integer = "%d?%d?%d",   -- 0-999 (but 0-255 for values)
+    --percent = "%d?%d?%d%%", -- 0-999% (but 0-100% for values)
+    float   = "%d?%.?%d+",  -- this is _very_ crude, but probably fast
+    comma   = "%s*,%s*",    -- whitespace optional comma
+    space   = "%s+",        -- whitespace required
+    slash   = "%s*/%s*",    -- whitespace optional slash
+    rgb     = "^rgb%(%s*",  -- rgb functional header
+    rgba    = "^rgba%(%s*", -- rgba functional header
+    hsl     = "^hsl%(%s*",  -- hsl functional header
+    hsla    = "^hsla%(%s*", -- hsla functional header
+    eol     = "%s*%)$",     -- functional EOL
+}
+-- capture a string
+local function capt(s)
+    return "(" .. s .. ")"
+end
+-- match string followed by %
+local function perc(s)
+    return s .. "%%"
+end
```

Comment:
The above two functions are meant to improve readability, right? I am not sure if it would be easier to just write the strings directly in the patterns, but I'll be quiet on this...

Reply:
Readability is correct. Since the integer and percent is the same, only adding the `%%` I could keep integer and capture that one but require the percent sign afterwards.